### PR TITLE
[MXNET-807] Support integer label type in ctc_loss operator

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
This PR fixed part of the issues in #10995. It supports integer type in label just as WarpCTC does. 
Also added a unit test to test large class issue fixed by another PR #11834

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Support integer label type
- [X] Created unit test for ctc_loss operator in CPU, GPU and integer labels

## Comments ##
- Also added a unit test for large arrays to cover the fix by another PR #11834

### CODE REVIEW

File: src/operator/contrib/ctc_loss-inl.h [hunk 1]
```diff
@@ -256,66 +255,69 @@ class CTCLossOp : public Operator {
...
+      std::vector<int> packed_labels;
+      std::vector<int> label_lengths(batch_size);
+
+      if (param_.use_label_lengths) {
+        int kLabelLength = 2 + param_.use_data_lengths;
+        exceed_cudnn_limit =
+          PackLabelByLength(labels, in_data[kLabelLength].get<xpu, 1, DType>(s),
+                           &packed_labels, &label_lengths);
+      } else {
+        exceed_cudnn_limit = LabelTensorToPackedVector(labels, param_.blank_label == 0?0:-1,
```
Comment (lebeg) [inline src/operator/contrib/ctc_loss-inl.h]: Some formatting issues with whitespaces and indentation

**Comment by lebeg** [comment]
@apeforest Could you rebase your change on latest master?

**Comment by apeforest** [comment]
@lebeg @samskalicky I have merged from master and the check now passed. Please review the PR again. Thanks

File: src/operator/contrib/ctc_loss-inl.h (hunk 1 above)
Comment (apeforest) [reply src/operator/contrib/ctc_loss-inl.h]: Sorry, what was exactly the issue? The make lint seems to pass.

Comment (lebeg) [reply src/operator/contrib/ctc_loss-inl.h]: Ping

**Comment by apeforest** [comment]
Hi @szha, @samskalicky asks if we could move this operator from contrib to mxnet regular. Do you have any suggestion? Thanks!

**Comment by szha** [comment]
@Jerryzcn @zhiheng-huang you're likely interested in this.

File: src/operator/contrib/ctc_loss-inl.h [hunk 2]
```diff
@@ -50,11 +50,10 @@
...
 namespace mxnet {
 namespace op {
-
 namespace ctc_loss {
-enum CTCLossOpInputs { kData, kLabel };
-enum CTCLossOpOutputs { kOut, kGrad };
-enum CTCLossOpForwardResource { kTempSpace };
+  enum CTCLossOpInputs { kData, kLabel };
+  enum CTCLossOpOutputs { kOut, kGrad };
+  enum CTCLossOpForwardResource { kTempSpace };
```
Comment (szha) [inline src/operator/contrib/ctc_loss-inl.h]: which case does this part fall into? how were we able to check it in without breaking master build if it's either of the cases?

File: src/operator/contrib/ctc_loss-inl.h [hunk 3]
```diff
@@ -331,21 +333,20 @@ class CTCLossOp : public Operator {
...
     Tensor<xpu, 3, real_t> data_grad =
-        in_grad[ctc_loss::kData].get<xpu, 3, real_t>(s);
+      in_grad[ctc_loss::kData].get<xpu, 3, real_t>(s);
     Tensor<xpu, 1, real_t> output_grad =
-        out_grad[ctc_loss::kOut].get<xpu, 1, real_t>(s);
+      out_grad[ctc_loss::kOut].get<xpu, 1, real_t>(s);
 
     Tensor<xpu, 3, real_t> data_grad_computed =
-        out_data[ctc_loss::kGrad].get<xpu, 3, real_t>(s);
+      out_data[ctc_loss::kGrad].get<xpu, 3, real_t>(s);
```
Comment (szha) [inline src/operator/contrib/ctc_loss-inl.h]: same here.

File: src/operator/contrib/ctc_loss-inl.h (hunk 2 above)
Comment (apeforest) [reply src/operator/contrib/ctc_loss-inl.h]: Sorry, overlooked these lines. I have removed them.

File: src/operator/contrib/ctc_loss-inl.h (hunk 3 above)
Comment (apeforest) [reply src/operator/contrib/ctc_loss-inl.h]: Removed. However, these 4 space indentation seems a violation of Google C++ style guide. Are we ignoring them in the lint?
https://google.github.io/styleguide/cppguide.html

File: tests/python/unittest/test_contrib_operator.py [hunk 4]
```diff
@@ -244,6 +244,64 @@ def assert_match(inputs, x, y, threshold, is_ascend=False):
     assert_match([[0.5, 0.6], [0.1, 0.2], [0.3, 0.4]], [1, -1, 0], [2, 0], 1e-12, False)
     assert_match([[0.5, 0.6], [0.1, 0.2], [0.3, 0.4]], [-1, 0, 1], [1, 2], 100, True)
 
+def test_ctc_loss_op():
```
[the remark this answers is not in the document]
Comment (apeforest) [reply tests/python/unittest/test_contrib_operator.py]: Thanks for the reference. Moving the unit test there.

File: tests/python/unittest/test_contrib_operator.py [hunk 5]
```diff
@@ -244,6 +244,64 @@ def assert_match(inputs, x, y, threshold, is_ascend=False):
...
+    x = np.random.uniform(size=(seq_len, batch_size, num_classes))
+    y = np.random.randint(0, num_classes, size=(batch_size, label_len))
+
+    def test_cpu(x, y):
+        data = mx.nd.array(x, ctx=mx.cpu(0))
+        label = mx.nd.array(y, ctx=mx.cpu(0))
+        loss = mx.nd.contrib.ctc_loss(data=data, label=label)
+        loss = mx.nd.make_loss(loss)
+        expected_output = [9.604521, 7.096151, 4.906869, 5.5237527, 5.9895644, 5.584548, 
+                           5.528411, 5.765914, 6.740701, 5.2625823]
```
[the remark this answers is not in the document]
Comment (apeforest) [reply tests/python/unittest/test_contrib_operator.py]: Did not notice the unit test in test_operator.py. I have removed this one.

**Comment by apeforest** [comment]
@szha This PR is to fix the unsupported integer label type. If we were to refactor this operator altogether, I would prefer to do it in another PR. What do you think? 

Also, as @samskalicky suggested, do you think it is mature to move this operator from contrib to regular? If that's the case, we can create another ticket to perform this migration together with refactoring.

File: tests/python/unittest/test_operator.py [hunk 6]
```diff
@@ -4516,6 +4516,27 @@ def test_ctc_loss():
...
+
+@with_seed(1)
+def test_ctc_loss_with_large_classes():
+    ctx = default_context()
+    batch_size = 1024
+    seq_len = 35
+    label_len = 10
+    num_classes = 6000
+    x = np.random.uniform(size=(seq_len, batch_size, num_classes))
+    y = np.random.randint(0, num_classes, size=(batch_size, label_len))
```
Comment (szha) [inline tests/python/unittest/test_operator.py]: again this does not seem like a good way of testing this.

Comment (apeforest) [reply tests/python/unittest/test_operator.py]: Any suggestion to test the large classes? I could compare this with WarpCtc implementation result if that can be treated as golden.

File: tests/python/unittest/test_operator.py [hunk 7]
```diff
@@ -4516,6 +4516,27 @@ def test_ctc_loss():
...
 
+    # Test 3: check use integer type as label
+    labels3 = np.array([[2, 3, 1], [2, 0, 0]], dtype=np.int32)
+    true_loss = np.array([7.3557, 5.4091], dtype=np.float32) # from Torch
+    check_ctc_loss(acts2, labels3, true_loss)
+
+@with_seed(1)
+def test_ctc_loss_with_large_classes():
+    ctx = default_context()
+    batch_size = 1024
```
Comment (szha) [inline tests/python/unittest/test_operator.py]: How does this help to verify the correctness?

File: tests/python/unittest/test_operator.py (hunk 6 above)
Comment (szha) [reply tests/python/unittest/test_operator.py]: Make a small example, calculate a the value and test for that, like in any other CTC tests. Since this is for testing the type, the batch size and sequence lengths are irrelevant.

File: tests/python/unittest/test_operator.py (hunk 7 above)
Comment (apeforest) [reply tests/python/unittest/test_operator.py]: I simply used the example reported in the original issue to make sure this fix addressed that.

Comment (szha) [reply tests/python/unittest/test_operator.py]: The issue that needs testing is the type of the labels, so a large batch size doesn't seem helpful or necessary for verifying the correctness.

Comment (szha) [reply tests/python/unittest/test_operator.py]: Also, tests with fixed seed are treated as a test quality issue and are being eliminated right now.

File: tests/python/unittest/test_operator.py (hunk 6 above)
Comment (apeforest) [reply tests/python/unittest/test_operator.py]: The label type is tested in line 4520. This testcase is to test the large number of classes that would crash reported in issue #10995

File: tests/python/unittest/test_operator.py (hunk 7 above)
Comment (apeforest) [reply tests/python/unittest/test_operator.py]: The label type is tested in line 4520. This testcase is to test the large number of classes that would crash reported in issue #10995

Comment (szha) [reply tests/python/unittest/test_operator.py]: OK, then make a test for it. Batch size is still not relevant, is it?

Comment (apeforest) [reply tests/python/unittest/test_operator.py]: It is not the batch_size in training. It is the size of the vocabulary. We need this variable to create the 3D tensor.

Comment (apeforest) [reply tests/python/unittest/test_operator.py]: updated the variable name and removed the fixed seed

Comment (szha) [reply tests/python/unittest/test_operator.py]: No, it really is not, the vocabulary size, regardless of how you name it. Please check the API doc and see its usage.

File: tests/python/unittest/test_operator.py [hunk 8]
```diff
@@ -4516,6 +4516,25 @@ def test_ctc_loss():
...
+    n = 35
+    l = 10
+    num_classes = 6000
+    x = np.random.uniform(size=(n, m, num_classes))
+    y = np.random.randint(0, num_classes, size=(m, l))
+ 
+    data = mx.nd.array(x, ctx=ctx)
+    label = mx.nd.array(y, ctx=ctx)
+    loss = mx.nd.contrib.ctc_loss(data=data, label=label)
+    assert loss.asnumpy().shape[0] == m
```
[the remark this answers is not in the document]
Comment (apeforest) [reply tests/python/unittest/test_operator.py]: Just to test the it does not crash

[the remark this answers is not in the document]
Comment (szha) [reply tests/python/unittest/test_operator.py]: This test does not crash on the master branch without the change either.

[the remark this answers is not in the document]
Comment (apeforest) [reply tests/python/unittest/test_operator.py]: That's true. This unit test is not to test my fix. It was to test an earlier PR #11834 which did not include a unit test but was merged somehow.

[the remark this answers is not in the document]
Comment (szha) [reply tests/python/unittest/test_operator.py]: Thanks for that. Still, the batch size is unnecessarily large. Why not make the test run faster? Also, there's still no test that covers the loss of precision problem that the integer label type solves, which is part of your fix. Would you mind adding that please?

File: tests/python/unittest/test_operator.py (hunk 7 above)
Comment (apeforest) [reply tests/python/unittest/test_operator.py]: Sorry for my misunderstanding the API. I have updated the unit tests based on your suggestion. Please review it again. Thanks!

File: tests/python/unittest/test_operator.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (apeforest) [reply tests/python/unittest/test_operator.py]: Updated the batch size to 2.

**Comment by apeforest** [comment]
Created a ticket https://issues.apache.org/jira/browse/MXNET-912 to move this operator from contrib to regular.

**Comment by apeforest** [comment]
ctc_loss operator yields different result in python3 and python2 environment breaking the newly added unit test. I am investigating the rootcause.
